### ISSUE TITLE
Supporting experimental decorators (not filtering out in rewrite-tsconfig)

### ISSUE THREAD

**Issue by AnsonT** [davidje13/neutrino-typescript#7]
Trying to use libraries like inversifyjs. While I've been able to enable metadata and decorators using babel plugins, ESLint still complains about the decorator because experimentalDecorators is filtered out when rewriting tsconfig

**Comment by davidje13** [comment]
In theory this is just a case of removing [compilerOptions.js#L30](https://github.com/davidje13/neutrino-typescript/blob/master/compilerOptions.js#L30)

How have you configured babel to work with it? (which plugins?) — I could make the `experimentalDecorators` flag automatically configure the same stuff (or at least have it check them for consistency). When they eventually become non-experimental I definitely want this package to do everything needed to make them work with babel; might as well start now.

I haven't used decorators much myself (except with Angular or NestJS where it's all handled internally); do you know if `emitDecoratorMetadata` is needed too? (I'm not clear on what it does, but maybe it's related to the generation of `.d.ts` files?)

As much as possible, I want it to be impossible to set tsconfig flags which are incompatible with the babel config.

**Comment by AnsonT** [comment]
```javascript
module.exports = {
  use: [
    typescript({ tsconfig: {
      compilerOptions: {
        strict: true,
        allowJs: true,
        declaration: true,
      },
    } }),
    typescriptLint(),
    eslint(),
    jest(),
    node({
      babel: {
        plugins: [
          "babel-plugin-transform-typescript-metadata",
          ["@babel/plugin-proposal-decorators", { "legacy": true }],
        ]
      }
    }),
  ],
};
```

With metadata, I also needed to add 'reflect-metadata' as a dependency and import 'reflect-metadata' in index.ts

**Comment by AnsonT** [comment]
Perhaps you can have 'experimentalDecorator: { force: true }' to allow overriding the safe options.

**Comment by davidje13** [comment]
OK, that setup looks pretty much like the recommendation from https://babeljs.io/docs/en/babel-plugin-transform-typescript

How about this:

If `@babel/plugin-proposal-decorators` is configured, `experimentalDecorators` is set to `true` automatically.
And if `babel-plugin-transform-typescript-metadata` is configured, `emitDecoratorMetadata` is set to `true` (still not sure if that's actually needed though, but probably doesn't hurt).

Both `experimentalDecorators` and `emitDecoratorMetadata` would remain unsettable in the regular `tsconfig` block, but they would be figured out automatically based on the presence of these plugins.

This matches the existing behaviour for supporting `jsx` (it implicitly enables jsx if `@babel/plugin-transform-react-jsx`, `@babel/plugin-syntax-jsx`, or `@babel/preset-react` are found).

Should be easy to implement. Any objections?

**Comment by AnsonT** [comment]
Metadata probably isn't needed, and experimentalDecorator is so that VS Code/ESLint doesn't complain. I don't have any objections with checking for @babel/plugin-proposal-decorators and setting the flag automatically, although it doesn't feel very discoverable.

An alternative is to add the @babel/plugin-proposal-decorators if experimentalDecorator is enabled in tsconfig. That may be easier to understand for a new user.

**Comment by davidje13** [comment]
> An alternative is to add the @babel/plugin-proposal-decorators if experimentalDecorator is enabled in tsconfig. That may be easier to understand for a new user.

Yeah that was my original thinking, but after looking into it a bit, I see that TypeScript's decorators are super legacy and probably going to change a whole lot when decorators officially hit Stage 3. For that reason, I'm thinking if people want to use them, they should have to do it themselves.

I pushed a change to master which seems to work (didn't test thoroughly but it writes the correct tsconfig.json). I'll cut a new release (1.1.5) in a moment.

For discoverability, I've added a custom error message if anybody tries to set those properties.
